contracts/MarginCalculator.sol

```diff
@@ -2,11 +2,279 @@
+pragma experimental ABIEncoderV2;
+
+import {Initializable} from "./packages/oz/upgradeability/Initializable.sol";
+import {SafeMath} from "./packages/oz/SafeMath.sol";
+import {OtokenInterface} from "./interfaces/OtokenInterface.sol";
+import {OracleInterface} from "./interfaces/OracleInterface.sol";
+import {AddressBookInterface} from "./interfaces/AddressBookInterface.sol";
+import {FixedPointInt256} from "./libs/FixedPointInt256.sol";
+import {MarginAccount} from "./libs/MarginAccount.sol";
 
 /**
  *
  */
-// solhint-disable-next-line no-empty-blocks
-contract MarginCalculator {
+contract MarginCalculator is Initializable {
+    using SafeMath for uint256;
+    using FixedPointInt256 for int256;
+    address public addressBook;
+
+    function init(address _addressBook) external initializer {
+        addressBook = _addressBook;
+    }
+
+    /**
+     * @notice Return the cash value of an expired oToken.
+     * @dev For call return = Max (0, ETH Price - oToken.strike)
+     * @dev For put return Max(0, oToken.strike - ETH Price)
+     * @param _otoken otoken address
+     * @return the cash value of an expired otoken
+     */
+    function getExpiredCashValue(address _otoken) public view returns (uint256) {
+        require(_otoken != address(0), "MarginCalculator: Invalid token address.");
+        OtokenInterface otoken = OtokenInterface(_otoken);
+        require(now > otoken.expiryTimestamp(), "MarginCalculator: Otoken not expired yet.");
+
+        uint256 strikePrice = otoken.strikePrice();
+        (uint256 underlyingPrice, bool isFinalized) = _getUnderlyingPrice(_otoken);
+        require(isFinalized, "MarginCalculator: Oracle price not finalized yet.");
+
+        if (otoken.isPut()) {
+            return strikePrice > underlyingPrice ? strikePrice.sub(underlyingPrice) : 0;
+        } else {
+            return underlyingPrice > strikePrice ? underlyingPrice.sub(strikePrice) : 0;
+        }
+    }
+
+    /**
+     * @notice returns the net value of a vault in the valid collateral asset for that vault i.e. USDC for puts/ ETH for calls
+     * @param _vault _vault the theoretical vault that needs to be checked
+     * @param _demonimated _denominated the token the result is denominated in. Must be the same as short.collateral for now.
+     * @return netValue the amount by which the margin is above or below the required amount.
+     * @return isExcess true if there's excess margin in the vault. In this case, collateral can be taken out from the vault. False if there is insufficient margin and additional collateral needs to be added to the vault to create the position.
+     */
+    function getExcessMargin(MarginAccount.Vault memory _vault, address _demonimated)
+        public
+        view
+        returns (uint256 netValue, bool isExcess)
+    {
+        // ensure the # asset in collateral, long and short array is valid.
+        _checkAssetCount(_vault);
+        // ensure the long asset is valid for the short asset.
```

Comment:
```suggestion
        // if there is a short option in the vault, ensure that the long option series being used is valid a margin.
```

Reply:
Same, I've add this to the function description.

Reply:
sure, even then, its nice to have that info here.

Reply:
If you can't understand the code without the comment, then the problem is with the  function naming. I think we should rename both functions instead of add duplicate comments.

What are better naming suggestions for `_checkAssetCount` and `_checkLongAsset`?

Reply:
_checkIsValidSpread
_checkIsValidNakedOption

Reply:
_isMaringableLong

Reply:
Updated!